### PR TITLE
Base64 Encoding simplification + optimization

### LINKED ISSUE #518: StoredMessage.Data as System.Buffers.Text
System.Buffers.Text [has To/From for utf8 byte sequences](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.text.base64.decodefromutf8?view=net-8.0) and it looks like they work with netstandard20, adapting to such an approach longer term may help consolidate this code as well as lower the deserialization cost of KV and other stream reads in NET8 (as we can hold the base64 in UTF8 bytes rather than UTF16 string, maybe pool in future too? Not sure how easy it would be to do that part though...)

see also https://github.com/nats-io/nats.net.v2/pull/513#discussion_r1642041779

### PR DESCRIPTION
**Pull request by to11mtm**
Resolves #518 

tl;dr - we are optimizing a lot of our base64 encode parsing by using `System.Text.Buffers` as it has NetStandard2.0 compatibility, which will also allow us to have less code to maintain long term. :)

This is a semi-breaking change however, as it does the following:

1. `System.Text.Json` is updated to the newest version, to support `ReadOnlyMemory<byte>`
   1. this version is compatible with NetStandard2.0 and Net6, so, win-win?
   2. Other win-win, by doing this, we instead rely on the implementations of how `System.Text.Json` handles `ReadOnlyMemory<byte>`, i.e. it does the Base64 decoding internally and should be 'good' for all versions (if there is issue we can always PR them!)
2. We use `ReadOnlyMemory<Byte>` for our model here now. If someone is somehow relying on that type for some test or other behavior it will be a breaking change
  1. FWIW this should be a -compile- breaking change, but if someone has a dis-proof of that I'm all ears.

### CODE REVIEW

File: src/NATS.Client.JetStream/NATS.Client.JetStream.csproj
Comment (to11mtm) [inline src/NATS.Client.JetStream/NATS.Client.JetStream.csproj:41]: Admittedly, IDK if I should be putting this elsewhere, but this was the 'lowest need' for the PR.

As noted, this works gr8 with netstandard2.0 and net6, so the biggest 'hazard' is consumers may have to deal with other changes.

File: src/NATS.Client.ObjectStore/NatsObjStore.cs
Comment (to11mtm) [inline src/NATS.Client.ObjectStore/NatsObjStore.cs:560]: Ironically, this optimizes NET8 too 😁. Not sure how much of a hot path this can be but why not?

That said, If there is one regret, IDK how to make this easily 'composable where this doesn't feel like a copypaste of the parse with buffer'. OTOH I had had a colleague spend some time teaching me go so the overton window for pragmatism has changed. 🤷

**Review by to11mtm — COMMENTED** [review]
Added some thoughts/reasoning to the changes given.

**Comment by to11mtm** [comment]
Thinking these errors -may- be related to some whitespace/etc present in the payload.

`Convert.FromBase64String` tends to be more tolerant of that... FWIW we should still be able to 'safely' trim things if that's the case (and I don't forsee a perf penalty since `FromBaase64String` is doing it too...)

Just gotta get my local test flow working again. 😅

**Comment by mtmk** [comment]
> `Convert.FromBase64String` tends to be more tolerant of that... FWIW we should still be able to 'safely' trim things if that's the case (and I don't forsee a perf penalty since `FromBaase64String` is doing it too...)

it's not only us we should be compatible with other clients as well, but I'd say that probably won't be a problem once we verified we're able to read and write proper base64. we should double check once local tests are passing (I can do that as well)

File: src/NATS.Client.ObjectStore/NatsObjStore.cs
Comment (to11mtm) [inline src/NATS.Client.ObjectStore/NatsObjStore.cs:545]: oops, should still clean this up.

**Comment by to11mtm** [comment]
> > `Convert.FromBase64String` tends to be more tolerant of that... FWIW we should still be able to 'safely' trim things if that's the case (and I don't forsee a perf penalty since `FromBaase64String` is doing it too...)
> 
> it's not only us we should be compatible with other clients as well, but I'd say that probably won't be a problem once we verified we're able to read and write proper base64. we should double check once local tests are passing (I can do that as well)

🤯 

I am amazed.

Tests were -actually- failing because `ReadOnlyMemory<byte>` in JSON Schema, will handle Base64 all on it's own! 🤯 

Once I was able to get tests (kinda) running locally it was pretty obvious to see what was happening. (will discuss to see if we should file something for the sake of doc or easing for others in future).

On one hand, we are semi 'at mercy' of however `System.Text.Json` is doing the Base64 decode, OTOH it's a lot less code for us to maintain and I'm guessing that we are still winning overall on alloc/perf since we are avoiding a UTF8->String conversion here.

**Comment by mtmk** [comment]
> Tests were -actually- failing because `ReadOnlyMemory<byte>` in JSON Schema, will handle Base64 all on it's own! 🤯

woow! very cool

**Comment by galvesribeiro** [comment]
> Tests were -actually- failing because ReadOnlyMemory<byte> in JSON Schema, will handle Base64 all on it's own! 

Yeah, there is a converter for it. Similar to `byte[]`.

> On one hand, we are semi 'at mercy' of however System.Text.Json is doing the Base64 decode, OTOH it's a lot less code for us to maintain and I'm guessing that we are still winning overall on alloc/perf since we are avoiding a UTF8->String conversion here.

Yeah, I would leave that to STJ. There is a huge work over the past couple of major releases to optimize the UTF8 handling in it. I don't see why we should need to roll our own in this case. 

What we should do is to make sure all types that come from Nats are using STJ Source Generators. That would avoid the reflection costs at runtime. Besides that, I won't see why would we go lower than that. I mean, yeah, we can always optimize something but it gets to a point of diminish return.

**Comment by to11mtm** [comment]
> Yeah, there is a converter for it. Similar to `byte[]`.

> Yeah, I would leave that to STJ. There is a huge work over the past couple of major releases to optimize the UTF8 handling in it. I don't see why we should need to roll our own in this case.

Agreed.

> What we should do is to make sure all types that come from Nats are using STJ Source Generators. That would avoid the reflection costs at runtime. Besides that, I won't see why would we go lower than that. I mean, yeah, we can always optimize something but it gets to a point of diminish return.

Agreed, if we do see an opportunity we can always contrib to upstream if needed. :)

I believe there's already sourcegen in play for STJ (I remember seeing some errors when I had bad composition), however your comment did lead me to look and I think we can also obviate a decent amount of code in ObjectStore with this change...

Although, not sure if we should keep that in same PR or separate, just for sake of sizing... open to opinions here.

**Comment by galvesribeiro** [comment]
> Although, not sure if we should keep that in same PR or separate, just for sake of sizing... open to opinions here.

Yeah, lets do on a separated round.

**Comment by to11mtm** [comment]
> > Although, not sure if we should keep that in same PR or separate, just for sake of sizing... open to opinions here.
> 
> Yeah, lets do on a separated round.

Created an issue to handle that (#557)

... TBH I think I'm good to pull this out of draft, my only concern is the weird flicker on unit tests. This latest push failed on test/dotnet (2.9) pull_request, the previous push failed on test/donet(latest).

I am no stranger to flickery CI but will defer to the opinions of core team here.

**Comment by galvesribeiro** [comment]
>  my only concern is the weird flicker on unit tests. 

When I was discussing the netstandard2.0 support, the tests flakiness were mentioned. I think it is essentially the fixtures as your test results are looking good. 

I suggested using TestContainers on it to get the fixtures working properly. I'll take a look on how to improve it and will create an issue.

**Review by mtmk — APPROVED** [review]
LGTM thanks @to11mtm

**Pull request merged**
